sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,256 @@
+{
+    ng_udp_hdr_t *hdr;
+    ng_pktsnip_t *udp_snip;
+    ng_netreg_entry_t *sendto;
+
+    /* get udp snip and hdr */
+    LL_SEARCH_SCALAR(pkt, udp_snip, type, NG_NETTYPE_UDP);
+    hdr = (ng_udp_hdr_t *)udp_snip->data;
+    /* fill in size field */
+    hdr->length = byteorder_htons(pkt->next->next->size + sizeof(ng_udp_hdr_t));
+
+    /* and forward packet to the network layer */
+    sendto = ng_netreg_lookup(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
+    /* throw away packet if no one is interested */
+    if (sendto == NULL) {
+        DEBUG("udp: cannot send packet because network layer not found\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    /* send packet to network layer */
+    ng_pktbuf_hold(pkt, ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL) - 1);
+    while (sendto != NULL) {
+        ng_netapi_send(sendto->pid, pkt);
+        sendto = ng_netreg_getnext(sendto);
+    }
+}
+
+void *_event_loop(void *arg)
+{
+    (void)arg;
+    msg_t msg;
+    msg_t msg_queue[NG_UDP_MSG_QUEUE_SIZE];
+
+    /* initialize message queue */
+    msg_init_queue(msg_queue, NG_UDP_MSG_QUEUE_SIZE);
+
+    /* dispatch NETAPI messages */
+    while (1) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case NG_NETAPI_MSG_TYPE_RCV:
+                DEBUG("udp: NG_NETAPI_MSG_TYPE_RCV\n");
+                _receive((ng_pktsnip_t *)msg.content.ptr);
+                break;
+            case NG_NETAPI_MSG_TYPE_SND:
+                DEBUG("udp: NG_NETAPI_MSG_TYPE_SND\n");
+                _send((ng_pktsnip_t *)msg.content.ptr);
+                break;
+            case NG_NETAPI_MSG_TYPE_SET:
+            case NG_NETAPI_MSG_TYPE_GET:
+                msg.type = NG_NETAPI_MSG_TYPE_ACK;
+                msg.content.value = (uint32_t)-ENOTSUP;
+                msg_reply(&msg, &msg);
+                break;
+            default:
+                DEBUG("udp: received unidentified message\n");
+                break;
+        }
+    }
+
+    /* never reached */
+    return NULL;
+}
+
+int ng_udp_calculate_csum(ng_pktsnip_t *hdr, ng_pktsnip_t *pseudo_hdr)
+{
+    uint16_t csum;
+    ng_udp_hdr_t *udp;
+
+    if (hdr == NULL || hdr->next == NULL) {
+        return -EFAULT;
+    }
+    if (hdr->type != NG_NETTYPE_UDP) {
+        return -EBADMSG;
+    }
+
+    csum = _calc_csum(hdr, pseudo_hdr, hdr->next);
+    if (csum == 0) {
+        return -ENOENT;
+    }
+    udp = (ng_udp_hdr_t *)hdr->data;
+    udp->checksum = byteorder_htons(csum);
+    return 0;
+}
+
+int ng_udp_hdr_build(pktsnip_t *payload,
+                     uint8_t *src, size_t src_len,
+                     uint8_t *dest, size_t dest_len) {
+    ng_pktsnip_t *res;
+    ng_udp_hdr_t *hdr;
+
+    /* allocate header */
+    res = ng_pktbuf_add(payload, NULL, sizeof(ng_udp_hdr_t), NG_NETTYPE_UDP);
+    if (res == NULL) {
+        return NULL;
+    }
+    /* initialize header */
+    hdr = (ng_udp_hdr_t *)res->data;
+    hdr->src_port = byteorder_htons(&((uint16_t *)src));
```

Comment:
Check if `src == NULL` or `src_len == sizeof(uint16_t)` before-hand, also you are giving `byteorder_htons()` a pointer of a `uint16_t` pointer. In anycase should be type-punning (this stuff: `*((uint16_t *)src)`) avoided, since it messes with  optimization, valgrind and static code analysis (and the compiler will throw a warning). I would propose to make `src` and `dst` `uint16_t` pointers to begin with and we modify `ng_netreg_hdr_build()` accordingly.